client/src/components/Explorer/reducers/nodes.js

```diff
@@ -19,8 +18,10 @@ const node = (state = defaultPageState, { type, payload }) => {
     return state || defaultPageState;
 
   case 'GET_PAGE_SUCCESS':
-    return Object.assign({}, state, payload.data, {
+    return Object.assign({}, state, {
       isError: false,
+      parent: (payload.data.meta.parent && payload.data.meta.parent.id) || null,
+      data: payload.data,
```

Comment:
It looks like `parent` is just derived from information that was already accessible in the reduced state. Could we avoid adding this computed data here, and instead retrieve the info as needed where it is needed?

Reply:
Mind if I leave this one until after the move to TypeScript?

Reply:
Hmm, I’m not sure. If you want to reduce refactoring work I would suggest to:

- Keep existing components using `pushPage` and `popPage` action creators, just redefine them to be based on the `GOTO_PAGE` action under the hood.
- Stop changing the shape of this `nodes` reducer – there is no need for the extra `data` structure if we just access the parent data from `page.meta.parent && page.meta.parent.id`.
- Could be even tidier by upgrading to a recent version of Babel so we could do `page.meta.parent?.id`

Reply:
> Stop changing the shape of this nodes reducer – there is no need for the extra data structure if we just access the parent data from page.meta.parent && 

While implementing https://github.com/wagtail/wagtail/pull/6474 I was reminded why I changed this. The nodes reducer mixes content from the API with local state. If we added a field to the API that clashed with a local one, weird stuff might happen.

Reply:
> Keep existing components using pushPage and popPage action creators, just redefine them to be based on the GOTO_PAGE action under the hood.

One issue with this is popPage doesn't have any arguments so there's no way of knowing what the destination page should be.

In the last, this was done by tracking all ancestors, but this gets tricky to do when we allow people to jump to any postition in the tree.